ReadMe.md
# Monal
## A cross platform, modern XMPP client for iOS and MacOS
![Build](https://travis-ci.org/anurodhp/Monal.svg?branch=develop)
[![Translation status](https://hosted.weblate.org/widgets/monal/-/svg-badge.svg)](https://hosted.weblate.org/engage/monal/?utm_source=widget)

This is the Monal XMPP client as found in the app store.  If you want to  use the latest stable versions, search for Monal in the iOS or OSX app store. Visit my blog to read about the development: www.monal.im

You can get the current releases in the App Store;

* [iOS](https://itunes.apple.com/us/app/monal-free-xmpp-chat/id317711500?mt=8)

* [Mac](https://apps.apple.com/app/id1499227291?mt=12)

### Impressions from the iOS version
<p float="left">
<img src ="https://monal.im/wp-content/uploads/2020/02/Simulator-Screen-Shot-iPhone-8-2020-02-17-at-15.58.26-1.png" width=280 >
<img src="https://monal.im/wp-content/uploads/2020/02/Simulator-Screen-Shot-iPhone-8-2020-02-17-at-15.56.17-4.png" width="280">  <img src="https://monal.im/wp-content/uploads/2020/02/Simulator-Screen-Shot-iPhone-8-2020-02-17-at-14.59.51-1.png" width="280">
</p>

### Impressions from the MacOS version

<img src="https://monal.im/wp-content/uploads/2020/02/Screen-Shot-2020-02-15-at-10.28.38-PM-1.png" width="600">

### Translations

We host and manage translations via [Weblate](https://hosted.weblate.org/engage/monal/).

### Platform information

Monal requires at least iOS 12 and macOS 10.15 (Catalina)

### Supported XEPs

Take a look at this list to get information on [supported XEPs by Monal](https://github.com/anurodhp/Monal/blob/develop/XEPsupport.md).

### License
Monal is licensed under the BSD license. Any code contributions should be compatible with that license.  ** NO GPL ** .  By contributing to this project, you agree that your code is not GPL or any similarly restrictive license. You agree that your code can be used to publish in App stores such as Apple's that use DRM.

### Pull Requests
I take pull requests. Please  use the develop branch to make changes. Please take a look at:

- [Overview: Tasks and Issues](https://github.com/anurodhp/Monal/issues/322) 
- [Support Monal](https://github.com/anurodhp/Monal/issues/363)

Monal is licensed under the BSD license. Any code contributions should be compatible with that license.  ** NO GPL ** .  By contributing to this project, you agree that your code is not GPL or any similarly restrictive license. You agree that your code can be used to publish in App stores such as Apple's that use DRM.
=======
[Install App](itms-services://?action=download-manifest&url=https://www.eightysoft.de/monal/monal.plist)
>>>>>>> bc8e5026... Configure travis build


changes.txt
commit 1b748d419f2e15a99a43d383da68bf5165dfa69f
Author: Friedrich Altheide <11352905+FriedrichAltheide@users.noreply.github.com>
Date:   Sat Feb 6 06:13:40 2021 +0100

    Localizations: Switching to monal-im source


scripts/build.sh
#!/bin/sh

cd Monal

# Abort on Error
set -e
export PING_SLEEP=10s
export BUILD_OUTPUT=../scripts/build.out
touch $BUILD_OUTPUT
dump_output() {
   echo Tailing the last 50 lines of output:
   tail -50 $BUILD_OUTPUT  
}
error_handler() {
  echo ERROR: An error was encountered with the build.
  dump_output
  exit 1
}
# If an error occurs, run our error handler to output a tail of the build
trap 'error_handler' ERR
# Set up a repeating loop to send some output to Travis.
bash -c "while true; do echo \$(date) - building ...; sleep $PING_SLEEP; done" &
PING_LOOP_PID=$!


ls -l ~/Library/MobileDevice/Provisioning\ Profiles/

security default-keychain -s ios-build.keychain
# Unlock the keychain
security unlock-keychain -p travis ios-build.keychain
# Set keychain timeout to 1 hour for long builds
security set-keychain-settings -t 3600 -l ~/Library/Keychains/ios-build.keychain

echo ""
echo "*********************************"
echo "*     Installing macOS Pods     *"
echo "*********************************"
pod install
pod update

echo ""
echo "***************************"
echo "*     Archiving macOS     *"
echo "***************************"
xcrun xcodebuild -workspace "Monal.xcworkspace" -scheme "Monal" -sdk macosx -configuration Debug -destination 'generic/platform=macOS,variant=Mac Catalyst,name=Any Mac' -archivePath "build/macos_$APP_NAME.xcarchive" clean archive CODE_SIGN_IDENTITY="$APP_DEVELOPER_NAME" CODE_SIGN_STYLE="Manual" GCC_PREPROCESSOR_DEFINITIONS='$GCC_PREPROCESSOR_DEFINITIONS IS_ALPHA=1' BUILD_LIBRARIES_FOR_DISTRIBUTION=YES SUPPORTS_MACCATALYST=YES >> $BUILD_OUTPUT 2>&1

echo ""
echo "****************************"
echo "*     Exporting macOS      *"
echo "****************************"
# see: https://gist.github.com/cocoaNib/502900f24846eb17bb29
# and: https://forums.developer.apple.com/thread/100065
# and: for developer-id distribution (distribution *outside* of appstore) an developer-id certificate must be used for building
xcodebuild -exportArchive -archivePath "build/macos_$APP_NAME.xcarchive" -exportPath "build/app" -exportOptionsPlist ../scripts/catalyst_exportOptions.plist
echo "build dir:"
ls -l "build"

echo ""
echo "**************************"
echo "*     Packing macOS      *"
echo "**************************"
cd build/app
mv "$APP_NAME.app" "$APP_NAME.alpha.app"
tar -cf "$APP_NAME.tar" "$APP_NAME.alpha.app"
cd ../..
ls -l build/app
	

echo ""
echo "*******************************"
echo "*     Installing iOS Pods     *"
echo "*******************************"
sed 's/###ios_only###//' Podfile >Podfile.new
mv Podfile Podfile.old
mv Podfile.new Podfile
pod install
pod update

echo ""
echo "*************************"
echo "*     Archiving iOS     *"
echo "*************************"
xcrun xcodebuild -workspace "Monal.xcworkspace" -scheme "Monal" -sdk iphoneos -configuration Debug -archivePath "build/ios_$APP_NAME.xcarchive" clean archive CODE_SIGN_IDENTITY="$IOS_DEVELOPER_NAME" CODE_SIGN_STYLE="Manual" GCC_PREPROCESSOR_DEFINITIONS='$GCC_PREPROCESSOR_DEFINITIONS IS_ALPHA=1' >> $BUILD_OUTPUT 2>&1

echo ""
echo "*************************"
echo "*     Exporting iOS     *"
echo "*************************"
# see: https://gist.github.com/cocoaNib/502900f24846eb17bb29
# and: https://forums.developer.apple.com/thread/100065
xcodebuild -exportArchive -archivePath "build/ios_$APP_NAME.xcarchive" -exportPath "build/ipa" -exportOptionsPlist ../scripts/exportOptions.plist
echo "build dir:"
ls -l "build"


# The build finished without returning an error so dump a tail of the output
#dump_output
# nicely terminate the ping output loop
kill $PING_LOOP_PID
exit


Monal/Classes/MLConstants.h
//
//  MLConstants.h
//  Monal
//
//  Created by Anurodh Pokharel on 7/13/13.
//
//

#import <Foundation/Foundation.h>
#import "MLHandler.h"

@import CocoaLumberjack;
static const DDLogLevel ddLogLevel = DDLogLevelVerbose;
#import "MLLogFileManager.h"
#import "MLLogFormatter.h"


//configure app group constants
#define kAppGroup @"group.monal"
#define kMonalKeychainName @"Monal"

//this is in seconds
#if TARGET_OS_MACCATALYST
	#define SHORT_PING 16.0
	#define LONG_PING 32.0
#else
	#define SHORT_PING 4.0
	#define LONG_PING 16.0
#endif

@class MLContact;

//some typedefs used throughout the project
typedef void (^contactCompletion)(MLContact *selectedContact);
typedef void (^accountCompletion)(NSInteger accountRow);
typedef void (^monal_void_block_t)(void);

typedef enum NotificationPrivacySettingOption {
    DisplayNameAndMessage,
    DisplayOnlyName,
    DisplayOnlyPlaceholder
} NotificationPrivacySettingOption;


//some useful macros
#define weakify(var) __weak __typeof__(var) AHKWeak_##var = var
#define strongify(var) _Pragma("clang diagnostic push") _Pragma("clang diagnostic ignored \"-Wshadow\"") __strong __typeof__(var) var = AHKWeak_##var; _Pragma("clang diagnostic pop")
#define nilWrapper(var) (var ? var : [NSNull null])
#define nilExtractor(var) (var == [NSNull null] ? nil : var)

//some xmpp related constants
#define kRegServer @"yax.im"
#define kMessageDeletedBody @"eu.siacs.conversations.message_deleted"

#define kXMLNS @"xmlns"
#define kId @"id"
#define kJid @"jid"
#define kMessageId @"kMessageId"

#define kRegisterNameSpace @"jabber:iq:register"
#define kDataNameSpace @"jabber:x:data"

//all other constants needed
#define kMonalNewMessageNotice @"kMLNewMessageNotice"
#define kMonalDeletedMessageNotice @"kMonalDeletedMessageNotice"
#define kMonalDisplayedMessageNotice @"kMonalDisplayedMessageNotice"
#define kMonalHistoryMessagesNotice @"kMonalHistoryMessagesNotice"
#define kMLMessageSentToContact @"kMLMessageSentToContact"
#define kMonalSentMessageNotice @"kMLSentMessageNotice"
#define kMonalMessageFiletransferUpdateNotice @"kMonalMessageFiletransferUpdateNotice"

#define kMonalLastInteractionUpdatedNotice @"kMonalLastInteractionUpdatedNotice"
#define kMonalMessageReceivedNotice @"kMonalMessageReceivedNotice"
#define kMonalMessageDisplayedNotice @"kMonalMessageDisplayedNotice"
#define kMonalMessageErrorNotice @"kMonalMessageErrorNotice"
#define kMonalReceivedMucInviteNotice @"kMonalReceivedMucInviteNotice"
#define kXMPPError @"kXMPPError"
#define kScheduleBackgroundFetchingTask @"kScheduleBackgroundFetchingTask"
#define kMonalUpdateUnread @"kMonalUpdateUnread"

#define kMLHasConnectedNotice @"kMLHasConnectedNotice"
#define kMonalFinishedCatchup @"kMonalFinishedCatchup"
#define kMonalFinishedOmemoBundleFetch @"kMonalFinishedOmemoBundleFetch"
#define kMonalUpdateBundleFetchStatus @"kMonalUpdateBundleFetchStatus"
#define kMonalIdle @"kMonalIdle"
#define kMonalFiletransfersIdle @"kMonalFiletransfersIdle"

#define kMonalPresentChat @"kMonalPresentChat"

#define kMLMAMPref @"kMLMAMPref"


#define kMonalCallStartedNotice @"kMonalCallStartedNotice"
#define kMonalCallRequestNotice @"kMonalCallRequestNotice"

#define kMonalAccountStatusChanged @"kMonalAccountStatusChanged"
#define kMonalAccountAuthRequest @"kMonalAccountAuthRequest"

#define kMonalRefresh @"kMonalRefresh"
#define kMonalContactRefresh @"kMonalContactRefresh"
#define kMonalXmppUserSoftWareVersionRefresh @"kMonalXmppUserSoftWareVersionRefresh"
#define kMonalBlockListRefresh @"kMonalBlockListRefresh"

// max count of char's in a single message (both: sending and receiving)
#define kMonalChatMaxAllowedTextLen 2048

#if TARGET_OS_MACCATALYST
#define kMonalChatFetchedMsgCnt 75
#else
#define kMonalChatFetchedMsgCnt 50
#endif

//contact cells
#define kusernameKey @"username"
#define kfullNameKey @"fullName"
#define kaccountNoKey @"accountNo"
#define kstateKey @"state"
#define kstatusKey @"status"

//info cells
#define kaccountNameKey @"accountName"
#define kinfoTypeKey @"type"
#define kinfoStatusKey @"status"

//blocking rules
#define kBlockingNoMatch 0
#define kBlockingMatchedNodeHostResource 1
#define kBlockingMatchedNodeHost 2
#define kBlockingMatchedHostResource 3
#define kBlockingMatchedHost 4

//temp not for release
#ifndef DEBUG
#define DEBUG 1
#endif

//use this to completely disable omemo in build
//#ifndef DISABLE_OMEMO
//#define DISABLE_OMEMO 1
//#endif

//build MLXMLNode query statistics (will only optimize MLXMLNode queries if *not* defined)
//#define QueryStatistics 1



Monal/Classes/MLLogInViewController.m
//
//  MLLogInViewController.m
//  Monal
//
//  Created by Anurodh Pokharel on 11/9/18.
//  Copyright © 2018 Monal.im. All rights reserved.
//

#import "MLLogInViewController.h"
#import "MBProgressHUD.h"
#import "DataLayer.h"
#import "MLXMPPManager.h"
#import "xmpp.h"

@import SAMKeychain;
@import QuartzCore;
@import SafariServices;

@class MLQRCodeScanner;

@interface MLLogInViewController ()

@property (nonatomic, strong) MBProgressHUD* loginHUD;
@property (nonatomic, weak) UITextField* activeField;
@property (nonatomic, strong) NSString* accountNo;

@end

@implementation MLLogInViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.topImage.layer.cornerRadius=5.0;
    self.topImage.clipsToBounds=YES;
}

- (void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
    [nc addObserver:self selector:@selector(connected:) name:kMonalFinishedCatchup object:nil];
#ifndef DISABLE_OMEMO
    [nc addObserver:self selector:@selector(updateBundleFetchStatus:) name:kMonalUpdateBundleFetchStatus object:nil];
#endif
    [nc addObserver:self selector:@selector(omemoBundleFetchFinished:) name:kMonalFinishedOmemoBundleFetch object:nil];
    [nc addObserver:self selector:@selector(error) name:kXMPPError object:nil];

    if(@available(iOS 13.0, *))
    {
        // nothing to do here
    }
    else
    {
        [self.qrScanButton setTitle:NSLocalizedString(@"QR", "MLLoginView: QR-Code Button iOS12 only") forState:UIControlStateNormal];
    }
    [self registerForKeyboardNotifications];
}

-(void) openLink:(NSString *) link
{
    NSURL *url= [NSURL URLWithString:link];
    
    if ([url.scheme isEqualToString:@"http"] || [url.scheme isEqualToString:@"https"] ) {
        SFSafariViewController *safariView = [[ SFSafariViewController alloc] initWithURL:url];
        [self presentViewController:safariView animated:YES completion:nil];
    }
}

-(IBAction) registerAccount:(id)sender;
{
    [self openLink:@"https://monal.im/welcome-to-xmpp/"];
}

-(IBAction) login:(id)sender
{
    [self login];
}

-(void) login
{
    self.loginHUD = [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    self.loginHUD.label.text = NSLocalizedString(@"Logging in", @"");
    self.loginHUD.mode=MBProgressHUDModeIndeterminate;
    self.loginHUD.removeFromSuperViewOnHide=YES;

    NSString* jid = self.jid.text;
    NSString* password = self.password.text;
    
    NSArray* elements = [jid componentsSeparatedByString:@"@"];

    NSString* domain;
    NSString* user;
    //if it is a JID
    if([elements count] > 1)
    {
        user = [elements objectAtIndex:0];
        domain = [elements objectAtIndex:1];
    }
   
    if(!user || !domain)
    {
        self.loginHUD.hidden = YES;
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Invalid Credentials", @"") message:NSLocalizedString(@"Your XMPP account should be in in the format user@domain. For special configurations, use manual setup.", @"") preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Close",@"") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [alert dismissViewControllerAnimated:YES completion:nil];
        }]];
        [self presentViewController:alert animated:YES completion:nil];
        return;
    }
    
    if(password.length == 0)
    {
        self.loginHUD.hidden = YES;
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Invalid Credentials",@"") message:NSLocalizedString(@"Please enter a password.",@"") preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Close",@"") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [alert dismissViewControllerAnimated:YES completion:nil];
        }]];
        [self presentViewController:alert animated:YES completion:nil];
        return;
    }
    
    if([[DataLayer sharedInstance] doesAccountExistUser:user.lowercaseString andDomain:domain.lowercaseString]) {
        self.loginHUD.hidden = YES;
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Duplicate Account", @"") message:NSLocalizedString(@"This account already exists on this instance", @"") preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Close", @"") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [alert dismissViewControllerAnimated:YES completion:nil];
        }]];
        [self presentViewController:alert animated:YES completion:nil];
        return;
    }

    NSMutableDictionary* dic  = [[NSMutableDictionary alloc] init];
    [dic setObject:domain.lowercaseString forKey:kDomain];
    [dic setObject:user.lowercaseString forKey:kUsername];
    [dic setObject:[HelperTools encodeRandomResource]  forKey:kResource];
    [dic setObject:@YES forKey:kEnabled];
    [dic setObject:@NO forKey:kDirectTLS];
    
    NSNumber* accountID = [[DataLayer sharedInstance] addAccountWithDictionary:dic];
    if(accountID) {
        self.accountNo = [NSString stringWithFormat:@"%@", accountID];
        [SAMKeychain setAccessibilityType:kSecAttrAccessibleAfterFirstUnlock];
        [SAMKeychain setPassword:password forService:kMonalKeychainName account:self.accountNo];
        [[MLXMPPManager sharedInstance] connectAccount:self.accountNo];
    }

    // open privacy settings
    if(![[HelperTools defaultsDB] boolForKey:@"HasSeenPrivacySettings"]) {
        [self performSegueWithIdentifier:@"showPrivacySettings" sender:self];
        return;
    }
}

-(void) connected:(NSNotification*) notification
{
    if([notification.userInfo[@"accountNo"] isEqualToString:self.accountNo])
    {
        [[HelperTools defaultsDB] setBool:YES forKey:@"HasSeenLogin"];
    #ifndef DISABLE_OMEMO
        dispatch_async(dispatch_get_main_queue(), ^{
            self.loginHUD.label.text = NSLocalizedString(@"Loading omemo bundles", @"");
        });
    #else
        [self kMonalFinishedOmemoBundleFetch];
    #endif
    }
}

#ifndef DISABLE_OMEMO
-(void) updateBundleFetchStatus:(NSNotification*) notification
{
    if([notification.userInfo[@"accountNo"] isEqualToString:self.accountNo])
    {
        dispatch_async(dispatch_get_main_queue(), ^{
            self.loginHUD.label.text = [NSString stringWithFormat:NSLocalizedString(@"Loading omemo bundles: %@ / %@", @""), notification.userInfo[@"completed"], notification.userInfo[@"all"]];
        });
    }
}
#endif

-(void) omemoBundleFetchFinished:(NSNotification*) notification
{
    if([notification.userInfo[@"accountNo"] isEqualToString:self.accountNo])
    {
        dispatch_async(dispatch_get_main_queue(), ^{
            self.loginHUD.hidden = YES;
            UIAlertController* alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Success!", @"") message:NSLocalizedString(@"You are set up and connected.", @"") preferredStyle:UIAlertControllerStyleAlert];
            [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Start Using Monal", @"") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                [self dismissViewControllerAnimated:YES completion:nil];
            }]];
            [self presentViewController:alert animated:YES completion:nil];
        });
    }
}

-(void) error
{
    dispatch_async(dispatch_get_main_queue(), ^{
        self.loginHUD.hidden=YES;
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Error", @"") message:NSLocalizedString(@"We were not able to connect your account. Please check your credentials and make sure you are connected to the internet.", @"") preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Close", @"") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [alert dismissViewControllerAnimated:YES completion:nil];
        }]];
        [self presentViewController:alert animated:YES completion:nil];
        
        if(self.accountNo)
            [[DataLayer sharedInstance] removeAccount:self.accountNo];
    });
}

-(IBAction) useWithoutAccount:(id)sender
{
    [[HelperTools defaultsDB] setBool:YES forKey:@"HasSeenLogin"];
    [self dismissViewControllerAnimated:YES completion:nil];
}

-(IBAction) tapAction:(id)sender
{
    [self.view endEditing:YES];
}



#pragma mark -textfield delegate

- (void)textFieldDidBeginEditing:(UITextField *)textField
{
    self.activeField = textField;
}

- (void)textFieldDidEndEditing:(UITextField *)textField
{
    self.activeField = nil;
}

// login on enter
-(void) enterPressed:(UIKeyCommand*)keyCommand
{
    [self login];
}

// List of custom hardware key commands
- (NSArray<UIKeyCommand *> *)keyCommands {
    return @[
        // enter
        [UIKeyCommand keyCommandWithInput:@"\r" modifierFlags:0 action:@selector(enterPressed:)],
    ];
}


#pragma mark - keyboard management

- (void)registerForKeyboardNotifications
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWasShown:)
                                                 name:UIKeyboardDidShowNotification object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillBeHidden:)
                                                 name:UIKeyboardWillHideNotification object:nil];
}

// Called when the UIKeyboardDidShowNotification is sent.
- (void)keyboardWasShown:(NSNotification*)aNotification
{
    NSDictionary* info = [aNotification userInfo];
    CGSize kbSize = [[info objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue].size;
    
    UIEdgeInsets contentInsets = UIEdgeInsetsMake(0.0, 0.0, kbSize.height, 0.0);
    self.scrollView.contentInset = contentInsets;
    self.scrollView.scrollIndicatorInsets = contentInsets;
    
    // If active text field is hidden by keyboard, scroll it so it's visible
    // Your app might not need or want this behavior.
    CGRect aRect = self.view.frame;
    aRect.size.height -= kbSize.height;
    if (!CGRectContainsPoint(aRect, self.activeField.frame.origin) ) {
        [self.scrollView scrollRectToVisible:self.activeField.frame animated:YES];
    }
}

// Called when the UIKeyboardWillHideNotification is sent
- (void)keyboardWillBeHidden:(NSNotification*)aNotification
{
    UIEdgeInsets contentInsets = UIEdgeInsetsZero;
    self.scrollView.contentInset = contentInsets;
    self.scrollView.scrollIndicatorInsets = contentInsets;
}

-(void) dealloc
{
    [self removeObservers];
}


-(void) removeObservers {
    NSNotificationCenter* nc = [NSNotificationCenter defaultCenter];
    [nc removeObserver:self];
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if([segue.identifier isEqualToString:@"scanQRCode"])
    {
        MLQRCodeScanner* qrCodeScanner = (MLQRCodeScanner*)segue.destinationViewController;
        qrCodeScanner.loginDelegate = self;
    }
    else if([segue.identifier isEqualToString:@"showPrivacySettings"])
    {
        // nothing todo
    }
    else
    {
        [self removeObservers];
    }
}

-(void) MLQRCodeAccountLoginScannedWithJid:(NSString*) jid password:(NSString*) password
{
    // Insert jid and password into text fields
    self.jid.text = jid;
    self.password.text = password;
    // Close QR-Code scanner
    [self.navigationController popViewControllerAnimated:YES];
}


@end


Monal/Classes/MLSettingsTableViewController.m
//
//  MLSettingsTableViewController.m
//  Monal
//
//  Created by Anurodh Pokharel on 12/26/17.
//  Copyright © 2017 Monal.im. All rights reserved.
//

#import "MLSettingsTableViewController.h"
#import "MLWebViewController.h"
@import SafariServices;

NS_ENUM(NSInteger, kSettingSection)
{
    kSettingSectionApp = 0,
    kSettingSectionSupport,
    kSettingSectionAbout,
    kSettingSectionCount
};

@interface MLSettingsTableViewController ()

@property (nonatomic, strong) NSArray* sections;
@property (nonatomic, strong) NSArray* appRows;
@property (nonatomic, strong) NSArray* supportRows;
@property (nonatomic, strong) NSArray* aboutRows;

@end

@implementation MLSettingsTableViewController 


- (IBAction)close:(id)sender
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.sections = @[NSLocalizedString(@"App", @""), NSLocalizedString(@"Support", @""), NSLocalizedString(@"About", @"")];
    
    self.appRows = @[
        NSLocalizedString(@"Quick Setup", @""),
        NSLocalizedString(@"Accounts",@""),
        NSLocalizedString(@"Privacy Settings",@""),
        NSLocalizedString(@"Notifications",@""),
        NSLocalizedString(@"Backgrounds",@""),
        NSLocalizedString(@"Sounds",@""),
        NSLocalizedString(@"Chat Logs",@"")
    ];
    self.supportRows = @[
        NSLocalizedString(@"Email Support", @""),
        NSLocalizedString(@"Submit A Bug", @"")
    ];

    self.aboutRows = @[
        NSLocalizedString(@"Rate Monal", @""),
        NSLocalizedString(@"Open Source", @""),
        NSLocalizedString(@"Privacy", @""),
        NSLocalizedString(@"About", @""),
#ifdef DEBUG
        NSLocalizedString(@"Log", @""),
#endif
        NSLocalizedString(@"Version", @"")
    ];

    self.splitViewController.preferredDisplayMode=UISplitViewControllerDisplayModeAllVisible;
#if !TARGET_OS_MACCATALYST
    if (@available(iOS 13.0, *)) {
        self.splitViewController.primaryBackgroundStyle=UISplitViewControllerBackgroundStyleSidebar;
    } else {
        // Fallback on earlier versions
    }
#endif
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return kSettingSectionCount;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    switch(section)
    {
        case kSettingSectionApp: return self.appRows.count;
        case kSettingSectionSupport: return self.supportRows.count;
        case kSettingSectionAbout: return self.aboutRows.count;
    }
    return 0;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:@"settingsCell" forIndexPath:indexPath];
    
    switch(indexPath.section)
    {
        case kSettingSectionApp: {
            cell.textLabel.text = self.appRows[indexPath.row];
            cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
            break;
        }
        case kSettingSectionSupport: {
            cell.textLabel.text = self.supportRows[indexPath.row];
            cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
            break;
        }
        case kSettingSectionAbout: {
            if(indexPath.row == (self.aboutRows.count - 1))
            {
                NSDictionary* infoDict = [[NSBundle mainBundle] infoDictionary];
                NSString* version = [infoDict objectForKey:@"CFBundleShortVersionString"];
                NSString* build = [infoDict objectForKey:@"CFBundleVersion"];
                
                cell.textLabel.text = [NSString stringWithFormat:NSLocalizedString(@"Version %@ (%@)", @""), version, build];
                cell.accessoryType = UITableViewCellAccessoryNone;
            } else {
                cell.textLabel.text = self.aboutRows[indexPath.row];
                cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
            }
            break;
        }
    }
    return cell;
}


-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return (section != kSettingSectionApp) ? self.sections[section] : 0;
}


-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    switch(indexPath.section)
    {
        case kSettingSectionApp: {
           
            switch ((indexPath.row)) {
                case 0:
                    [self performSegueWithIdentifier:@"showLogin" sender:self];
                    break;

                case 1:
                    [self performSegueWithIdentifier:@"showAccounts" sender:self];
                    break;

                case 2:
                    [self performSegueWithIdentifier:@"showPrivacySettings" sender:self];
                    break;

                case 3:
                    [self performSegueWithIdentifier:@"showNotification" sender:self];
                    break;

                case 4:
                    [self performSegueWithIdentifier:@"showBackgrounds" sender:self];
                    break;

                case 5:
                    [self performSegueWithIdentifier:@"showSounds" sender:self];
                    break;

                case 6:
                    [self performSegueWithIdentifier:@"showChatLog" sender:self];
                    break;

                default:
                    break;
            }
            break;
        }
        case kSettingSectionSupport: {
            switch ((indexPath.row)) {
                case 0:
                    [self composeMail];
                    break;
                    
                case 1:
                     [self openLink:@"https://github.com/anurodhp/Monal/issues"];
                    break;
                default:
                    break;
            }
            break;
        }
        case kSettingSectionAbout: {
            switch ((indexPath.row)) {
                case 0:
                    [self openStoreProductViewControllerWithITunesItemIdentifier:317711500];
                    break;
                    
                case 1:
                    [self performSegueWithIdentifier:@"showOpenSource" sender:self];
                    break;
                    
                case 2:
                    [self openLink:@"https://monal.im/privacy-policy/"];
                    break;
                    
                case 3:
                    [self openLink:@"https://monal.im/about/"];
                    break;
                    
                case 4:
                    [self performSegueWithIdentifier:@"showLogs" sender:self];
                    break;
               
                default:
                    break;
            }
            break;
        }
    }
}

-(void) openLink:(NSString *) link
{
    NSURL* url = [NSURL URLWithString:link];
    
    if ([url.scheme isEqualToString:@"http"] || [url.scheme isEqualToString:@"https"]) {
        SFSafariViewController *safariView = [[ SFSafariViewController alloc] initWithURL:url];
        [self presentViewController:safariView animated:YES completion:nil];
    }
}

#pragma mark - Actions


- (void)openStoreProductViewControllerWithITunesItemIdentifier:(NSInteger)iTunesItemIdentifier {
    SKStoreProductViewController *storeViewController = [[SKStoreProductViewController alloc] init];
    
    storeViewController.delegate = self;
    
    NSNumber* identifier = [NSNumber numberWithInteger:iTunesItemIdentifier];
    //, @"action":@"write-review"
    NSDictionary* parameters = @{ SKStoreProductParameterITunesItemIdentifier:identifier};
    
    [storeViewController loadProductWithParameters:parameters
                                   completionBlock:^(BOOL result, NSError *error) {
                                       if (result)
                                           [self presentViewController:storeViewController
                                                              animated:YES
                                                            completion:nil];
                                       else NSLog(@"SKStoreProductViewController: %@", error);
                                   }];
    
    
}

-(void)composeMail
{
    if([MFMailComposeViewController canSendMail]) {
        MFMailComposeViewController* composeVC = [[MFMailComposeViewController alloc] init];
        composeVC.mailComposeDelegate = self;
        [composeVC setToRecipients:@[@"info@monal.im"]];
        [self presentViewController:composeVC animated:YES completion:nil];
    }
    else  {
        UIAlertController* messageAlert =[UIAlertController alertControllerWithTitle:NSLocalizedString(@"Error", @"") message:NSLocalizedString(@"There is no configured email account. Please email info@monal.im .", @"") preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction* closeAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"Close", @"") style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
        }];
        [messageAlert addAction:closeAction];
        
        [self presentViewController:messageAlert animated:YES completion:nil];
    }
    
}

#pragma mark - Message ui delegate
- (void)mailComposeController:(MFMailComposeViewController *)controller
          didFinishWithResult:(MFMailComposeResult)result
                        error:(NSError *)error
{
    [controller dismissViewControllerAnimated:YES completion:nil];
}


#pragma mark - SKStoreProductViewControllerDelegate

-(void)productViewControllerDidFinish:(SKStoreProductViewController *)viewController {
    [viewController dismissViewControllerAnimated:YES completion:nil];
}


-(void)  prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if([segue.identifier isEqualToString:@"showOpenSource"])
    {
        UINavigationController *nav = (UINavigationController *)  segue.destinationViewController;
        MLWebViewController *web = (MLWebViewController *) nav.topViewController;
        
        NSBundle* mainBundle = [NSBundle mainBundle];
        NSString* myFile = [mainBundle pathForResource: @"opensource" ofType: @"html"];
    
        web.urltoLoad=[NSURL fileURLWithPath:myFile];
    }
}

@end
